obkvrpc/compressor.go

```diff
@@ -0,0 +1,156 @@
+type Lz4Decompressor struct{}
+type SnappyDecompressor struct{}
+type ZlibDecompressor struct{}
+type ZStdDecompressor struct{}
+
+func (d *NoneDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	return src, nil
+}
+
+func (d *Lz4Decompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	srcBytes := src.Bytes()
+	dstBytes := make([]byte, originLen)
+	n, err := lz4.UncompressBlock(srcBytes, dstBytes)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("Lz4Decompressor: length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
+	}
+	return bytes.NewBuffer(dstBytes), nil
+}
+
+func (d *SnappyDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	srcBytes := src.Bytes()
+	dstBytes, err := snappy.Decode(nil, srcBytes)
+	if err != nil {
+		return nil, err
+	}
+	if int32(len(dstBytes)) != originLen {
+		return nil, errors.New(fmt.Sprintf("SnappyDecompressor length of decompress content is not expected: actual-%d, expected-%d", int32(len(dstBytes)), originLen))
+	}
+	return bytes.NewBuffer(dstBytes), nil
+}
+
+func (d *ZlibDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	// 1. create zlib reader
+	zlibReader, err := zlib.NewReader(src)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func(zlibReader io.ReadCloser) {
+		err := zlibReader.Close()
+		if err != nil {
+			log.Warn(fmt.Sprintf("fail to close the zlibReader, errMsg: %s", err.Error()))
+		}
+	}(zlibReader)
+
+	// 2. create buffer
+	dstBuffer := new(bytes.Buffer)
+	// 3. write the decompress data to dstBuffer
+	n, err := io.Copy(dstBuffer, zlibReader)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("ZlibDecompressor:length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
+	}
+	return dstBuffer, nil
+}
+
+func (d *ZStdDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	zstdReader, err := zstd.NewReader(src)
+	if err != nil {
+		return nil, err
+	}
+	defer zstdReader.Close()
+	dstBuffer := new(bytes.Buffer)
+	n, err := io.Copy(dstBuffer, zstdReader)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("ZStdDecompressor: length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
+	}
+	return dstBuffer, nil
+}
+
+func getDecompressor(compressType protocol.ObCompressType) (DeCompressor, error) {
+	switch compressType {
+	case protocol.ObCompressTypeLZ4:
+		return &Lz4Decompressor{}, nil
+	case protocol.ObCompressTypeSnappy:
+		return &SnappyDecompressor{}, nil
+	case protocol.ObCompressTypeZlib:
+		return &ZlibDecompressor{}, nil
+	case protocol.ObCompressTypeZstd:
+		return &ZStdDecompressor{}, nil
+	case protocol.ObCompressTypeZstd_138:
+		return &ZStdDecompressor{}, nil
```

Comment:
Will ObCompressTypeZstd_138 and ObCompressTypeZstd use the same decompressor and be processed the same way?

Reply:
Yes. These two compress types are supported by observer rpc framework. They are the two different version of the same compress algorithms. It has been verified that can use the same decompressor to process in client and test case has been added in `query_with_compress_test.go`.